### PR TITLE
move some App CR reconciliation parts into sidecar

### PR DESCRIPTION
**Pull request by cppforlife**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -31,6 +31,7 @@ RUN curl -sLo cue.tgz https://github.com/cue-lang/cue/releases/download/v0.4.2/c
 COPY . .
 # helpful ldflags reference: https://www.digitalocean.com/community/tutorials/using-ldflags-to-set-version-information-for-go-applications
 RUN CGO_ENABLED=0 GOOS=linux go build -mod=vendor -ldflags="-X 'main.Version=$KCTRL_VER'" -trimpath -o controller ./cmd/main.go
+RUN CGO_ENABLED=0 GOOS=linux go build -mod=vendor -ldflags="-X 'main.Version=$KCTRL_VER'" -trimpath -o controller-sidecarexec ./pkg/sidecarexec/cmd/main.go
```
Comment (benmoss) [inline Dockerfile]: can we clean up toplevel `cmd` to have `cmd/controller/main.go` and `cmd/sidecar/main.go` or whatever? having mains under pkg seems odd

File: cmd/controller/run.go [hunk 2]
```diff
@@ -147,15 +151,24 @@ func Run(opts Options, runLog logr.Logger) error {
 		if err != nil {
 			return fmt.Errorf("Setting up Config reconciler watches: %s", err)
 		}
+
+		// Reconcile once synchronously to ensure configuration is loaded before any execution happens
```
Comment (benmoss) [inline cmd/controller/run.go]: I don't understand this comment, why do we have to do this now?

File: cmd/controller/run.go [hunk 3]
```diff
@@ -237,17 +250,3 @@ func Run(opts Options, runLog logr.Logger) error {
 
 	return nil
 }
-
-func logProxies(runLog logr.Logger) {
```
Comment (benmoss) [inline cmd/controller/run.go:241]: Unrelated change?

File: cmd/main.go [hunk 4]
```diff
@@ -40,16 +36,13 @@ func main() {
...
+	if err != nil {
+		mainLog.Error(err, "Exited run with error")
+		os.Exit(1)
+		panic("unreachable: controller exited 1")
 	}
 
-	controllerinit.Run(os.Args[0], os.Args[1:], log.WithName("init"))
-	panic("unreachable: init proc returned")
+	os.Exit(0)
+	panic("unreachable: controller exited 0")
```
Comment (benmoss) [inline cmd/main.go]: what is with this unreachable code stuff and the Exit(0) call?

File: config/deployment.yml [hunk 5]
```diff
@@ -49,6 +51,31 @@ spec:
...
+            memory: 100Mi
+        env:
+        - name: KAPPCTRL_SIDECAREXEC_SOCK
+          value: /etc/kappctrl-mem-tmp/sidecarexec.sock
+        volumeMounts:
+        - name: template-fs
+          mountPath: /etc/kappctrl-mem-tmp
+        - name: home
+          mountPath: /home/kapp-controller
+        #! mount over service account token to make it not available
```
Comment (benmoss) [inline config/deployment.yml:76]: interesting hack

File: pkg/config/config.go [hunk 6]
```diff
@@ -7,56 +7,30 @@ import (
 	"context"
```
Comment (benmoss) [inline pkg/config/config.go:7]: Why did all this need to change rather than just being called by the sidecar instead of the controller?

**Review by benmoss — COMMENTED** [review]
It's an interesting idea, and a step in the right direction at least in terms of isolating the control plane from the semi-user-invoked CLIs.

File: Dockerfile (hunk 1 above)
Comment (cppforlife) [reply Dockerfile]: yeah, absolutely. didnt want to push it in this PR, but happy to. or as a followup.

File: cmd/controller/run.go (hunk 2 above)
Comment (cppforlife) [reply cmd/controller/run.go]: since sidecar can be configured with ca certs / proxies we want to ensure that that happens before actual execution. any suggestions on rephrasing this comment?

File: cmd/controller/run.go (hunk 3 above)
Comment (cppforlife) [reply cmd/controller/run.go:241]: that got moved into sidecar itself since only sidecar needs to be configured with proxies.

File: cmd/main.go (hunk 4 above)
Comment (cppforlife) [reply cmd/main.go]: yeah... thats just weirdness... im ok removing it.

File: config/deployment.yml (hunk 5 above)
Comment (cppforlife) [reply config/deployment.yml:76]: the only way i could find how to "remove" sa token from one of the containers. this same trick is used by agones for game servers.

File: pkg/config/config.go (hunk 6 above)
Comment (cppforlife) [reply pkg/config/config.go:7]: sidecar doesnt have access to k8s client. it only exposes api how to configure ca cert / proxies. Config object got simplified to just be a thing that has info, vs thing that knows how to apply it.

File: config/deployment.yml (hunk 5 above)
Comment (praveenrewar) [reply config/deployment.yml:76]: 🤯

File: cmd/main.go (hunk 4 above)
Comment (cppforlife) [reply cmd/main.go]: removed.

File: cmd/controller/run.go (hunk 2 above)
Comment (cppforlife) [reply cmd/controller/run.go]: updated comment with a little bit more info.

File: Dockerfile (hunk 1 above)
Comment (cppforlife) [reply Dockerfile]: did it in this PR.

**Review by benmoss — APPROVED** [review]
